src/csharp/Grpc.Tools/implementation_notes.md

```diff
@@ -0,0 +1,219 @@
+
+Details of how NuGet packages can add custom build targets and properties to a project is documented
+here: [MSBuild .props and .targets in a package](https://learn.microsoft.com/en-us/nuget/concepts/msbuild-props-and-targets)
+
+Basically the `.props` and `.targets` files are automatically included in the projects - the `.props` at the top
+of the project and the `.targets` are added to the bottom of the project.
+
+## Visual Studio property pages
+
+For Visual Studio integration - these files provide the properties pages:
+
+* `build\_protobuf\Protobuf.CSharp.xml` (included from `Google.Protobuf.Tools.targets`)
+* `build\_grpc\Grpc.CSharp.xml` (included from `_Grpc.Tools.targets`)
+
+## Custom tasks DLLs
+
+DLLs containing the custom tasks are in:
+
+* `build\_protobuf\netstandard1.3`
+* `build\_protobuf\net45`
+
+## Protobuf compiler and C# gRPC plugin binaries
+
+Native binary executables for the protobuf compiler (_protoc_) and C# gRPC plugin (_grpc_csharp_plugin_) are
+included in the NuGet package. Included are binaries for various OSes (Windows, Linux, macOS) and
+CPU architectures (x86, x64, arm64).
+
+The build determines which executables to use for the particular machine that the it is being run on.
+These can be overridden by specifying MSBuild properties or environment variables to give the paths to custom executables:
+
+* `Protobuf_ProtocFullPath` property or `PROTOBUF_PROTOC` environment variable \
+Full path of protoc executable
+* `gRPC_PluginFullPath` property or `GRPC_PROTOC_PLUGIN` environment variable \
+Full path of gRPC C# plugin
+
+# Grpc.Tools custom build targets
+
+## Hooking the custom targets into the project build
+
+The custom targets hook into various places in a normal MSBuild build by specifying
+before/after targets at the relevant places. See
+[msbuild-targets](https://learn.microsoft.com/en-us/visualstudio/msbuild/msbuild-targets)
+for predefined targets.
+
+* Before `PrepareForBuild`
+    * we add `Protobuf_SanityCheck` that checks this is a supported project type, e.g. a C# project.
+* Before `BeforeCompile`
+    * we add all the targets that compile the `.proto` files and generate the expected `.cs` files. These files are added to those that get compiled by the C# compiler.
+* After `CoreClean`
+    * we add `Protobuf_Clean` that cleans the files generated by the protobuf compiler.
+
+## Custom tasks
+
+There are a few custom tasks needed by the targets. These are implemented in C# in the Grpc.Tools project
+and packaged in the file `Protobuf.MSBuild.dll` in the NuGet package.
+
+* ProtoToolsPlatform
+    * Works out the operating system and CPU architecture
+* ProtoCompilerOutputs
+    * Tries to work out the names and paths of the files that would be generated by the protobuf compiler and returns these as a list of _items_
+    * Also returns list of _items_ that is the same as the _Protobuf_ items passed in with the output directory metadata updated
```

Comment:
was `_items_` and `_Protobuf_` formatting intended or you forgot to update them? at least _Protobuf_ should be `Protobuf`?